src/sinks/util/service/net/unix.rs

```diff
@@ -0,0 +1,293 @@
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct UnixConnector {
+    path: PathBuf,
+    mode: UnixMode,
+    send_buffer_size: Option<usize>,
+}
+
+impl UnixConnector {
+    async fn connect(&self) -> Result<(PathBuf, UnixEither), UnixError> {
+        let either_socket = match self.mode {
+            UnixMode::Datagram => UnixDatagram::unbound()
+                .context(FailedToBindSnafu)
+                .and_then(|datagram| {
+                    datagram
+                        .connect(&self.path)
+                        .context(FailedToConnectSnafu)
+                        .map(|_| UnixEither::Datagram(datagram))
+                })?,
+            UnixMode::Stream => UnixStream::connect(&self.path)
+                .await
+                .context(FailedToConnectSnafu)
+                .map(UnixEither::Stream)?,
+        };
+
+        if let Some(send_buffer_size) = self.send_buffer_size {
+            if let Err(error) = net::set_send_buffer_size(&either_socket, send_buffer_size) {
+                warn!(%error, "Failed configuring send buffer size on Unix socket.");
+            }
+        }
+
+        Ok((self.path.clone(), either_socket))
+    }
+
+    async fn connect_backoff(&self) -> UnixEither {
+        // TODO: Make this configurable.
+        let mut backoff = ExponentialBackoff::from_millis(2)
+            .factor(250)
+            .max_delay(Duration::from_secs(60));
+
+        loop {
+            match self.connect().await {
+                Ok((path, either_socket)) => {
+                    emit!(UnixSocketConnectionEstablished { path: &path });
+                    return either_socket;
+                }
+                Err(error) => {
+                    emit!(UnixSocketOutgoingConnectionError { error });
+                    sleep(backoff.next().unwrap()).await;
+                }
+            }
+        }
+    }
+
+    /// Gets a `Healthcheck` based on the configured destination of this connector.
+    pub fn healthcheck(&self) -> Healthcheck {
+        let connector = self.clone();
+        Box::pin(async move { connector.connect().await.map(|_| ()).map_err(Into::into) })
+    }
+
+    /// Gets a `Service` suitable for sending data to the configured destination of this connector.
+    pub fn service(&self) -> UnixService {
+        UnixService::new(self.clone())
+    }
+}
+
+enum UnixServiceState {
+    /// The service is currently disconnected.
+    Disconnected,
+
+    /// The service is currently attempting to connect to the endpoint.
+    Connecting(BoxFuture<'static, UnixEither>),
+
+    /// The service is connected and idle.
+    Connected(UnixEither),
+
+    /// The service has an in-flight send to the socket.
+    ///
+    /// If the socket experiences an unrecoverable error during the send, `None` will be returned
+    /// over the channel to signal the need to establish a new connection rather than reusing the
+    /// existing connection.
+    Sending(oneshot::Receiver<Option<UnixEither>>),
+}
```

Comment:
Seems like the `<Service>State` enums from all this new service code are the same except a single type. Possible to extract with a generic?

Reply:
They could likely be boiled down to a single state type, parameterized over the connection object, yes.

Reply:
Resolved via the `NetworkConnector` stuff.